ir/src/main/scala/org/scalajs/ir/Types.scala

```diff
@@ -163,17 +163,38 @@ object Types {
     }
   }
 
+  sealed abstract class NonArrayTypeRef extends TypeRef
+
+  /** Primitive type reference. */
+  final case class PrimRef private[ir] (tpe: Type) extends NonArrayTypeRef {
+    override def equals(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
+
+    override def hashCode(): Int = super.hashCode()
+  }
+
+  final val VoidRef = PrimRef(NoType)
+  final val BooleanRef = PrimRef(BooleanType)
+  final val CharRef = PrimRef(CharType)
+  final val ByteRef = PrimRef(ByteType)
+  final val ShortRef = PrimRef(ShortType)
+  final val IntRef = PrimRef(IntType)
+  final val LongRef = PrimRef(LongType)
+  final val FloatRef = PrimRef(FloatType)
+  final val DoubleRef = PrimRef(DoubleType)
+  final val NullRef = PrimRef(NullType)
+  final val NothingRef = PrimRef(NothingType)
+
```

Comment:
Actually: Any reason these are not case objects? We could avoid the unchecked pattern matches then, no? Also it would be more consistent with the primitive types.

Reply:
Yeah, I thought a lot about that. I really wanted to make them case objects, but if I do that I cannot make `PrimRef` a case class, and it really should be a case class for pattern matching of the form `case PrimRef(tpe)` to be efficient. So I didn't really have a choice, unfortunately.

We could do an abominable thing and make `PrimRef` and `abstract case class`, while making the individuals (non-case) `object`s ... But if I do that I won't be able to argue with Martin that abstract case class are an abomination anymore.

Reply:
Right. And if you simply define an extractor, you do not get exhaustiveness checking.

But, now I can of course come back to a slight variation of `PrimType`: A sub hierarchy (`RefablePrimType`) would allow exhaustiveness checking on the inner type. The `final val`s could be kept as mere aliases or not.

In the same swoop, I'd avoid overriding `equals` and `hashCode`: It feels like premature optimization traded off for some safety.

WDYT?

Reply:
Yes, that's probably reasonable. I'll try that later today when I get the chance.

Reply:
Done.